src/validate.ts

```diff
@@ -74,65 +71,72 @@ export function validate(intentRequest: object, apiResponse: object, syncData?:
-             const trait = syncDevices[j]['traits'];
-             if (trait in TRAITS_COMMANDS_PAIR){
-               const validateQueryTraitStates = responseValidation(states, COMMAND_STATES_EXPECT[TRAITS_COMMANDS_PAIR[trait]]);
-                 if (validateQueryTraitStates) {
-                   return queryErrors.push(...validateQueryTraitStates);
-                 }
-             }
+          const trait = syncDevices[j]['traits'];
+          if (trait in TRAITS_COMMANDS_PAIR) {
+            const validateQueryTraitStates = responseValidation(states, COMMAND_STATES_EXPECT[TRAITS_COMMANDS_PAIR[trait]]);
+            if (validateQueryTraitStates) {
+              queryErrors.push(...validateQueryTraitStates);
+            }
           }
         }
       }
+    }
     return queryErrors.length ? queryErrors : undefined;
   } else if (responseType === 'action.devices.EXECUTE') {
     // validate with states schema
     const executeErrors : object[] = [];
     // gets the execution array from the intent request
     const execution = intentRequest['inputs'][0]['payload']['execution'];
-    const executionLength = intentRequest['inputs'][0]['payload']['execution'].length;
+    const executionLength = execution.length;
+
+    const validateExecuteAPI = responseValidation(apiResponse, EXECUTE_RESPONSE_SCHEMA);
+
+    if (validateExecuteAPI) {
+      executeErrors.push(...validateExecuteAPI);
+      return executeErrors;
+    }
 
     // identifies the part of the api response to validate against a schema
     const commands = apiResponse['payload']['commands'];
     const commandsLength = commands.length;
-    
-   const validateExecuteAPI = responseValidation(apiResponse, EXECUTE_RESPONSE_SCHEMA);
-   
-   if (validateExecuteAPI === undefined){
-        for (let i = 0; i < executionLength; i++) {
-          // gets the specific command
-          const commandName = execution[i]['command'];
-          if (commandName in COMMAND_STATES_EXPECT) {
-            for (let j = 0; j < commandsLength; j++) {
-              const states = commands[j]['states'];
-              const validateExecTraitStates = responseValidation(states, COMMAND_STATES_EXPECT[commandName]);
-              if (validateExecTraitStates) {
-                executeErrors.push(...validateExecTraitStates);
-              } else {
-                return validateExecuteAPI;
-              }
-            }
+    for (let i = 0; i < executionLength; i++) {
+      // gets the specific command
+      const commandName = execution[i]['command'];
+      if (commandName in COMMAND_STATES_EXPECT) {
+        for (let j = 0; j < commandsLength; j++) {
+          const states = commands[j]['states'];
+          const validateExecTraitStates = responseValidation(states, COMMAND_STATES_EXPECT[commandName]);
+          if (validateExecTraitStates) {
+            executeErrors.push(...validateExecTraitStates);
+          } else {
+            return validateExecuteAPI;
```

Comment:
I think we MIGHT be in the same scenario here as with attributes (i.e. we shouldn't return early just because one command validates)?

Reply:
True, probably best to delete the return line. Made these updates.